### DISCUSSION TITLE
3D surface grid with X,Y, and Z points

### DISCUSSION THREAD

**Discussion by gunjan71011** [pyvista/pyvista#2643] [Q&A]
Hello, I am trying to figure out how I be able to apply X,Y,Z points to create a 3-D surface grid using pyVista? Similar to the quick example, "Creating a Structured Surface" on the website (https://docs.pyvista.org/examples/00-load/create-structured-surface.html#sphx-glr-examples-00-load-create-structured-surface-py). But in my case, I have thousands of  X,Y,Z points in space. I am able to do this using matplotlib.pyplot in python however, I cannot seem to work out how to do it using pyVista.

**Comment by banesullivan** [comment]
Hi @gunjan71011! Could you perhaps share a data file with this type of point data set?

Are your points scattered? or do they have some sort of inherit structure?

Perhaps a quick/dirty approach would be to run a Delaunay triangulation much like [this example](https://docs.pyvista.org/examples/00-load/create-tri-surface.html). To say anything more, I'd need to see the dataset either with a data file or if you could take screenshots from plotting the point set with PyVista.

**Comment by banesullivan** [comment]
@gunjan71011, did you attach the file to the email when you replied? You may need to visit this thread’s webpage and drag’n’drop the file into the text box

**Comment by banesullivan** [comment]
```py
import pandas as pd
import numpy as np
import pyvista as pv

# Load Excel sheet using Pandas
# Note - you may need to `pip install xlrd`
df = pd.read_excel('/Users/bane/Downloads/WFA.xlsx')
coords = df[['X', 'Y', 'Z']].values

# Preview the coordinates to see what we are dealing with
poly = pv.PolyData(coords)
poly['order'] = np.linspace(0,1,poly.n_points)
poly.plot(notebook=False, show_grid=True, point_size=1)
```

<img width="580" alt="Screen Shot 2019-06-18 at 9 57 50 AM" src="https://user-images.githubusercontent.com/22067021/59699883-91077480-91af-11e9-97ae-bf47be419978.png">



The coordinates definitely have some sort of inherit structure present as I can see a structured grid of the points that is a bit rotated from the cartesian reference frame. This dataset would work well if we know the dimensions of the structured grid (nx by ny by 1) as we could recover that structure and add connectivity with quad cells, but this is incredibly difficult to do without knowing the dimensions of the grid (and assuming the coordinates are ordered which they appear to be from the linear trends in the color mapping shown above). Perhaps the original source of this dataset has more details on the grid's structure?

The quick and dirty method of Delaunay triangulation would work to create a surface, however this is a bit expensive for this dataset as it will have twice as many cells as a structured grid with quads and not have exactly the resolution you might desire:

```py
surf = poly.delaunay_2d()
surf.plot(notebook=False, show_grid=True)
```

<img width="580" alt="Screen Shot 2019-06-18 at 9 55 08 AM" src="https://user-images.githubusercontent.com/22067021/59699650-2f470a80-91af-11e9-95e5-a9c39df02d61.png">

<img width="624" alt="Screen Shot 2019-06-18 at 9 58 52 AM" src="https://user-images.githubusercontent.com/22067021/59699958-b300f700-91af-11e9-921c-9016438dc5d0.png">



So this new surface mesh has 38056 triangle cells and 19305 points. Ideally, I'd like to see this mesh as a structured grid with quads and not a polygonal surface, so here's an attempt to recover the structure of the gird which I did by brought force guessing dimensions that looked right and were multiples of the total number of points. I ended up with `[143, 135, 1]`:

```py
# Make the structured surface manually
structured = pv.StructuredGrid()
# Set coordinates
structured.points = coords
# Set the dimensions of the structured grid
structured.dimensions = [143, 135, 1]

# Apply an Elevation filter
elevation = structured.elevation()
elevation.plot(show_edges=True, show_grid=True, notebook=False)
```

<img width="624" alt="Screen Shot 2019-06-18 at 10 13 33 AM" src="https://user-images.githubusercontent.com/22067021/59701038-eba1d000-91b1-11e9-956e-ecaee2939221.png">

<img width="624" alt="Screen Shot 2019-06-18 at 10 13 26 AM" src="https://user-images.githubusercontent.com/22067021/59701053-efcded80-91b1-11e9-8676-7e841c4a7b95.png">

And there you have it, a properly structured surface from your point set with 19028 quad cells and 19305 points/nodes!

**Comment by banesullivan** [comment]
So in brief, the code to make the surface is:

```py
import pandas as pd
import pyvista as pv

# Load Excel sheet using Pandas
# Note - you may need to `pip install xlrd`
df = pd.read_excel('/Users/bane/Downloads/WFA.xlsx')
coords = df[['X', 'Y', 'Z']].values

# Make the structured surface manually
structured = pv.StructuredGrid()
# Set coordinates
structured.points = coords
# Set the dimensions of the structured grid
structured.dimensions = [143, 135, 1]

# Apply an Elevation filter
elevation = structured.elevation()
elevation.plot(show_edges=True, show_grid=True, notebook=False)
```

**Comment by akaszynski** [comment]
We should add this as an example in the docs.

Looks great @banesullivan!

**Comment by banesullivan** [comment]
Added a simple example and linked to this issue for a better/more complicated example

**Comment by gunjan71011** [comment]
Thanks that looks good, although being able to see the changes in elevation would be nice. But thanks for the help!

**Comment by banesullivan** [comment]
> although being able to see the changes in elevation would be nice

Maybe try adding exaggeration to the render view:

```py
plotter = pv.Plotter(notebook=False)
plotter.add_mesh(elevation)
plotter.set_scale(1, 1, 10)
plotter.show()
```

**Comment by gunjan71011** [comment]
Thanks Bane! That really makes the surface stand out. However, I noticed that if there are multiple surfaces in one plotter then all of it would be exaggerted by the same amount (plotter.set_scale(1, 1, 10)). Is there a way to give each mesh its own exaggeration?

**Comment by banesullivan** [comment]
> However, I noticed that if there are multiple surfaces in one plotter then all of it would be exaggerted by the same amount (plotter.set_scale(1, 1, 10)). Is there a way to give each mesh its own exaggeration?

You could give each dataset it's own exaggeration by editing the coordinates of each mesh individually and leaving the scale alone - the `set_scale` function exaggerates the entire render view's spatial reference system so it would be impossible for us to set a scale on each dataset individually and have them still appear next to each other. As a user, though, you could manually scale and shift datasets to your liking. Try editing and shifting the `.points` of your different meshes before plotting and not using the `set_scale` function:

```py
import pyvista as pv
from pyvista import examples

data = examples.download_bolt_nut()
bolt = data['bolt'].threshold(1)
nut = data['nut'].threshold(1)

# Original data
p = pv.Plotter()
p.add_mesh(bolt)
p.add_mesh(nut)
p.show(use_panel=False)
```
![download](https://user-images.githubusercontent.com/22067021/59791664-d9965f00-928f-11e9-9888-07b44feb4c42.png)

```py
# Scale and shift the bolt
bolt.points[:,1] *=3
bolt.points[:,1] -= bolt.points[:,1].ptp()/2

#scale and shift the nut
nut.points[:,0] *=3
nut.points[:,0] -= nut.points[:,0].ptp()/2
nut.points[:,2] *=3
nut.points[:,2] -= nut.points[:,2].ptp()/2

p = pv.Plotter()
p.add_mesh(bolt)
p.add_mesh(nut)
p.show(use_panel=False)
```
![download](https://user-images.githubusercontent.com/22067021/59791677-e2873080-928f-11e9-81a5-3c31e091d5db.png)

**Comment by akaszynski** [comment]
Would it be possible to implement either a set_scale on the actor or maybe make it a keyword arg for add_mesh?

**Comment by banesullivan** [comment]
That could also work... but you might still need to shift the datasets around to make sure they appear next to each other?

**Comment by akaszynski** [comment]
Good point.  Maybe a scene-wise implementation is best.

**Comment by banesullivan** [comment]
> Maybe a scene-wise implementation is best.

That's what I went with by implementing `set_scale` a while back. I very often have to exaggerate the Z axis and found that a scene wide scaling worked best - its quite rare (in my experience) to need individual exaggerations per mesh in a single scene
